### ISSUE TITLE
more on shutdown with mwi

### ISSUE THREAD

**Issue by juha-h** [alfredh/baresip#533]
I did some more tests.  This time I have two accounts (test and foo) with mwi.  When I quit, UN-REGISTER of one UA (test) happens too fast (before NOTIFY has arrived):
````
baresip is ready.
test@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b04 (x86_64/linux))) [1 binding]
mwi: subscribing to messages for sip:test@test.tutpro.com
presence: ua=0x561126a856e0 got event 1 (REGISTER_OK)
foo@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b04 (x86_64/linux))) [1 binding]
All 2 useragents registered successfully! (22 ms)
mwi: subscribing to messages for sip:foo@test.tutpro.com
presence: ua=0x561126a89890 got event 1 (REGISTER_OK)
----- MWI for sip:test@test.tutpro.com -----
Messages-Waiting: yes
Message-Account: sip:test@vm.test.tutpro.com
Voice-Message: 1/0 (0/0)

----- MWI for sip:foo@test.tutpro.com -----
Messages-Waiting: yes
Message-Account: sip:foo@vm.test.tutpro.com
Voice-Message: 2/1 (0/0)

Quit
ua: stop all (forced=0)
mwi: shutdown
presence: ua=0x561126a856e0 got event 4 (SHUTDOWN)
presence: subscriber: closing 0 subs
mwi: shutdown
presence: ua=0x561126a856e0 got event 3 (UNREGISTERING)
presence: ua=0x561126a89890 got event 3 (UNREGISTERING)
presence: ua=0x561126a89890 got event 4 (SHUTDOWN)
presence: subscriber: closing 0 subs
ua: ext_ref -> cannot unload mods
presence: ua=(nil) got event 5 (EXIT)
ua: sip-stack exit
ua exited -- stopping main runloop
module: unloading app vidloop
module: unloading app presence
module: unloading app mwi
module: unloading app menu
menu: close (redial current_attempts=0)
module: unloading app debug_cmd
module: unloading app contact
module: unloading app auloop
main: unloading modules..
 [   zrtp main]: DESTROYING LIBZRTP...
 [   zrtp main]: DESTROYING LIBZRTP - DONE
````
Here is log from my SIP Proxy:
````
root@char:/usr/src/orig/re# Sep 14 14:50:15 char /usr/bin/sip-proxy[9458]: INFO: Routing in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:test@test.tutpro.com>
Sep 14 14:50:15 char /usr/bin/sip-proxy[9458]: INFO: Routing in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:foo@test.tutpro.com>
Sep 14 14:50:15 char /usr/bin/sip-proxy[9458]: INFO: REGISTER <sip:test@test.tutpro.com> by <test@test.tutpro.com> from <192.168.43.107> using tcp is authorized
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: INFO: Handling in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:test@test.tutpro.com> to <sip:test@test.tutpro.com>
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: INFO: Routing locally generated NOTIFY <sip:test@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> to <sip:test@test.tutpro.com> from <sip:test@test.tutpro.com>
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: INFO: presence [notify.c:1614]: send_notify_request(): NOTIFY sip:test@test.tutpro.com via sip:127.0.0.1:5070;transport=tcp;r2=on;lr on behalf of sip:test@test.tutpro.com for event message-summary : 1574df7655be707a
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: INFO: Handling in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:foo@test.tutpro.com> to <sip:foo@test.tutpro.com>
Sep 14 14:50:15 char /usr/bin/sip-proxy[9460]: INFO: In-dialog NOTIFY <sip:test@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> is not found
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: INFO: Routing locally generated NOTIFY <sip:foo@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> to <sip:foo@test.tutpro.com> from <sip:foo@test.tutpro.com>
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: INFO: presence [notify.c:1614]: send_notify_request(): NOTIFY sip:foo@test.tutpro.com via sip:127.0.0.1:5070;transport=tcp;r2=on;lr on behalf of sip:foo@test.tutpro.com for event message-summary : 7caeb092adce4c47
Sep 14 14:50:15 char /usr/bin/pres-serv[10751]: ERROR: presence [subscribe.c:497]: delete_subs(): Failed to delete subscription from memory [slot: 233 ev: sip:test@test.tutpro.com pu: message-summary ci: 1574df7655be707a ft: 7bb33293947a5964 tt: 48b5b8952e8d0bca6256941910840e90-cd08]
Sep 14 14:50:15 char /usr/bin/sip-proxy[9460]: INFO: Routing in-dialog NOTIFY <sip:foo-0x561126a89890@192.168.43.107:41097;transport=tcp> from <sip:foo@test.tutpro.com> to <sip:192.168.43.107:37014;transport=tcp> based on gruu
Sep 14 14:50:15 char /usr/bin/sip-proxy[9458]: INFO: REGISTER <sip:foo@test.tutpro.com> by <foo@test.tutpro.com> from <192.168.43.107> using tcp is authorized
````
As you see from the above, foo is OK, i.e., it is UN-REGISTERed after NOTIFY has been received.

**Comment by juha-h** [comment]
I did the above test many times and in about one test in four or five, UN-REGISTER of test happens before UN-SUBSCRIBE NOTIFY has been processed.  It would be nice if baresip would do the right thing every time.

**Comment by juha-h** [comment]
juha-h writes:

> I did the above test many times and in about one test in four or five,
> UN-REGISTER of test happens before UN-SUBSCRIBE NOTIFY has been
> processed.  It would be nice if baresip would do the right thing every
> time.

I did more tests with "Using baresip as a library" example and having
the same two accounts (test and foo).  Same issue with that, i.e., a
couple of quits work fine and then failure.

Here is an example of failure:
````
test@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
mwi: subscribing to messages for sip:test@test.tutpro.com
foo@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [2 bindings]
ua event: REGISTER_OK
All 2 useragents registered successfully! (13 ms)
mwi: subscribing to messages for sip:foo@test.tutpro.com
ua event: MWI_NOTIFY
----- MWI for sip:test@test.tutpro.com -----
Messages-Waiting: yes
Message-Account: sip:test@vm.test.tutpro.com
Voice-Message: 1/0 (0/0)

ua event: MWI_NOTIFY
----- MWI for sip:foo@test.tutpro.com -----
Messages-Waiting: no
Message-Account: sip:foo@vm.test.tutpro.com
Voice-Message: 0/2 (0/0)
````
````
Quit
ua: stop all (forced=0)
ua event: SHUTDOWN
mwi: shutdown
presence: subscriber: closing 0 subs
ua event: SHUTDOWN
mwi: shutdown
ua event: UNREGISTERING
ua event: UNREGISTERING
presence: subscriber: closing 0 subs
ua: in use by app module, deferring module unloading

****** about 5 second break

ua event: EXIT
ua exited -- stopping main runloop
main: unloading modules..
````

Below is the messages from my proxy at Quit.  As you see, baresip
un-SUBs both UAs, but un-REGs them too early (before NOTIFYs have
arrived), which means that the proxy does not find the UAs anymore when
is tries to deliver the NOTIFYs.

I'll label this issue as a bug.

Question:  which timer fires after about 5 secs that causes the EXIT
event?  I would like to use that also in my Android app to properly stop
baresip.

-- Juha

````
Sep 19 16:52:02 char /usr/bin/sip-proxy[18269]: INFO: Routing in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:test@test.tutpro.com>
Sep 19 16:52:02 char /usr/bin/sip-proxy[18269]: INFO: Routing in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:foo@test.tutpro.com>
Sep 19 16:52:02 char /usr/bin/sip-proxy[18269]: INFO: REGISTER <sip:test@test.tutpro.com> by <test@test.tutpro.com> from <192.168.43.107> using tcp is authorized
Sep 19 16:52:02 char /usr/bin/pres-serv[10751]: INFO: Handling in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:test@test.tutpro.com> to <sip:test@test.tutpro.com>
Sep 19 16:52:02 char /usr/bin/sip-proxy[18269]: INFO: REGISTER <sip:foo@test.tutpro.com> by <foo@test.tutpro.com> from <192.168.43.107> using tcp is authorized
Sep 19 16:52:02 char /usr/bin/pres-serv[10751]: INFO: Routing locally generated NOTIFY <sip:test@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> to <sip:test@test.tutpro.com> from <sip:test@test.tutpro.com>
Sep 19 16:52:02 char /usr/bin/pres-serv[10751]: INFO: presence [notify.c:1614]: send_notify_request(): NOTIFY sip:test@test.tutpro.com via sip:127.0.0.1:5070;transport=tcp;r2=on;lr on behalf of sip:test@test.tutpro.com for event message-summary : 645e3e03ab241fb9
Sep 19 16:52:02 char /usr/bin/pres-serv[10751]: INFO: Handling in-dialog SUBSCRIBE <sip:127.0.0.1:5080;transport=tcp> from <sip:foo@test.tutpro.com> to <sip:foo@test.tutpro.com>
Sep 19 16:52:02 char /usr/bin/sip-proxy[18276]: INFO: In-dialog NOTIFY <sip:test@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> is not found
Sep 19 16:52:02 char /usr/bin/pres-serv[10751]: INFO: Routing locally generated NOTIFY <sip:foo@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> to <sip:foo@test.tutpro.com> from <sip:foo@test.tutpro.com>
Sep 19 16:52:02 char /usr/bin/pres-serv[10751]: INFO: presence [notify.c:1614]: send_notify_request(): NOTIFY sip:foo@test.tutpro.com via sip:127.0.0.1:5070;transport=tcp;r2=on;lr on behalf of sip:foo@test.tutpro.com for event message-summary : f27a001049ba7a8f
Sep 19 16:52:02 char /usr/bin/sip-proxy[18276]: INFO: In-dialog NOTIFY <sip:foo@test.tutpro.com;gr=urn:uuid:9b21c1a8-dc58-a3d6-83d5-7be67cbe70ce> is not found
````

**Comment by juha-h** [comment]
After commit https://github.com/alfredh/baresip/commit/c7a570b2df3f4fe4b00a187448b70da15eb942d3 baresip now quits fine when none or all of the two UAs have mwi enabled, but quit hangs if one has mwi enabled and one has not.  The behavior is the same with baresip app and with libbaresip example app.  The tests below have been made with libbaresip example app.  Source is master with some added/modified info() calls.

Both UAs have mwi disabled:
````
test@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
foo@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
All 2 useragents registered successfully! (146 ms)
Quit
ua: stop all (forced=0)
ua event: SHUTDOWN
mwi: ua event shutdown for sip:test@test.tutpro.com
ua event: SHUTDOWN
mwi: ua event shutdown for sip:foo@test.tutpro.com
Stopping 2 useragents.. 
ua event: UNREGISTERING
ua event: UNREGISTERING
ua event: EXIT
ua exited -- stopping main runloop
main: ua_close
main: conf_close
main: baresip_close
main: unloading modules..
````
Both UAs have mwi enabled:
````
test@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
mwi: subscribing to messages for sip:test@test.tutpro.com
foo@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
All 2 useragents registered successfully! (13 ms)
mwi: subscribing to messages for sip:foo@test.tutpro.com
mwi: notify for sip:test@test.tutpro.com
ua event: MWI_NOTIFY
----- MWI for sip:test@test.tutpro.com -----
Messages-Waiting: yes
Message-Account: sip:test@vm.test.tutpro.com
Voice-Message: 1/0 (0/0)

mwi: notify for sip:foo@test.tutpro.com
ua event: MWI_NOTIFY
----- MWI for sip:foo@test.tutpro.com -----
Messages-Waiting: no
Message-Account: sip:foo@vm.test.tutpro.com
Voice-Message: 0/2 (0/0)

Quit
ua: stop all (forced=0)
ua event: SHUTDOWN
mwi: ua event shutdown for sip:test@test.tutpro.com
mwi: shutting down sip:test@test.tutpro.com
ua event: SHUTDOWN
mwi: ua event shutdown for sip:foo@test.tutpro.com
mwi: shutting down sip:foo@test.tutpro.com
ua: in use by app module, deferring module unloading
mwi: dereffing sip:test@test.tutpro.com at deref_handler
ua event: UNREGISTERING
mwi: dereffing sip:foo@test.tutpro.com at deref_handler
ua event: UNREGISTERING
ua event: EXIT
ua exited -- stopping main runloop
main: ua_close
main: conf_close
main: baresip_close
main: unloading modules..
````
One UA (test) has mwi enabled and one (foo) disabled:
````
test@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
mwi: subscribing to messages for sip:test@test.tutpro.com
foo@test.tutpro.com: {1/TCP/v4} 200 OK (OpenSIPg SIP Proxy (5.2.0-0b05 (x86_64/linux))) [1 binding]
ua event: REGISTER_OK
All 2 useragents registered successfully! (14 ms)
mwi: notify for sip:test@test.tutpro.com
ua event: MWI_NOTIFY
----- MWI for sip:test@test.tutpro.com -----
Messages-Waiting: yes
Message-Account: sip:test@vm.test.tutpro.com
Voice-Message: 1/0 (0/0)

Quit
ua: stop all (forced=0)
ua event: SHUTDOWN
mwi: ua event shutdown for sip:test@test.tutpro.com
mwi: shutting down sip:test@test.tutpro.com
ua event: SHUTDOWN
mwi: ua event shutdown for sip:foo@test.tutpro.com
ua: in use by app module, deferring module unloading
mwi: dereffing sip:test@test.tutpro.com at deref_handler
ua event: UNREGISTERING
******* This UNREGISTERING event is for test. foo is never UNREGISTERed and baresip hangs
******* here forever unless I press q again:
Quit
ua: stop all (forced=0)
ua event: SHUTDOWN
mwi: ua event shutdown for sip:foo@test.tutpro.com
Stopping 1 useragent.. 
ua event: UNREGISTERING
ua event: EXIT
ua exited -- stopping main runloop
main: ua_close
main: conf_close
main: baresip_close
main: unloading modules..
````
I have been reading the source code for a few hours, and have not been able to figure out, why baresip does not UNREGISTER the ua that does not have mwi enabled.

Any help to solve this would be highly appreciated.

**Comment by juha-h** [comment]
I got quit working also in the mixed mwi case by uncommenting return in ua_stop_all() in the case when an app module holds a ref to a UA:
````
	if (ext_ref) {
		info("ua: in use by app module, deferring module unloading\n");
//		return;
	}
	else {
		module_app_unload();
	}
````
That allows ua_stop_all() to to stop those UAs that do not hold the refs.

Does this change have some negative side effects or can it be committed?

**Comment by alfredh** [comment]
one possible solution to this, is to check the nrefs to struct ua when shutting
down ua.c

for example, if mwi is loaded/used, the ref count will be 2 or more.

in this case, we can skip the unregister step, wait for 1-2 seconds, check again etc.

**Comment by juha-h** [comment]
Yes, but just commenting the return, every case that I have tested works fine.  defer_handler in mwi.c is launched 500 m after SHUTDOWN event and removes mwi->ua reference.  That will then eventually cause ua nrefs to go to zero and the ua is stopped.  What is wrong with this simple solution?

**Comment by alfredh** [comment]
the background for this logic, is related to modules and callback handlers
implemented in modules.

it is not possible to unload some application modules syncrononusly,
as they will call mem_deref on things like sipevent subscribe, and the callback handler
will be called asynronously from libre into the module later (after it was unloaded).
this will lead to a crash.

so we deploy a strategy to fix this; when the application is shutting down
send an event UA_SHUTDOWN to all event listeners. the app module receiving this
can now safely start its shutdown procedure. when the procedure is done it will
finally call `mem_deref(ua)` and the UA and the register client is gone.

I dont have a final solution just yet, but here is a couple of improvement points.

in the mwi module there is no need to iterate over all the uas, one can simply
look up the mwi object from the ua pointer:

```
	else if (ev == UA_EVENT_SHUTDOWN) {

		struct mwi *mwi = mwi_find(ua);
		struct le *le;

		if (mwi) {
			mwi->shutdown = true;

			if (mwi->sub) {
				mwi->sub = mem_deref(mwi->sub);
				tmr_start(&mwi->tmr, 500, deref_handler, mwi);
			}
			else
				mem_deref(mwi);
		}
	}
```


in `ua_stop_all` it should be possible to deref a ua, if the ref-count is 1.
this means that no-one else has a ref to it.

```c
	/* check if someone else has grabbed a ref to ua */
	le = uag.ual.head;
	while (le) {

		struct ua *ua = le->data;
		le = le->next;

		ua_event(ua, UA_EVENT_SHUTDOWN, NULL, NULL);

		if (mem_nrefs(ua) > 1) {

			list_unlink(&ua->le);
			list_flush(&ua->calls);
			mem_deref(ua);

			++ext_ref;
		}
		else
			mem_deref(ua);
	}
```

(NOTE: it only makes sense to sende UA_EVENT_SHUTDOWN if the ua is not destructed).

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> in the mwi module there is no need to iterate over all the uas, one can simply
> look up the mwi object from the ua pointer:

Yes, since all mwi UAs will get the shutdown event.

> in `ua_stop_all` it should be possible to deref a ua, if the ref-count is 1.
> this means that no-one else has a ref to it.
> 
```
> 	/* check if someone else has grabbed a ref to ua */
> 	le = uag.ual.head;
> 	while (le) {
> 
> 		struct ua *ua = le->data;
> 		le = le->next;
> 
> 		ua_event(ua, UA_EVENT_SHUTDOWN, NULL, NULL);
> 
> 		if (mem_nrefs(ua) > 1) {
> 
> 			list_unlink(&ua->le);
> 			list_flush(&ua->calls);
> 			mem_deref(ua);
> 
> 			++ext_ref;
> 		}
> 		else
> 			mem_deref(ua);
> 	}
```

The above is pretty much equivalent to my solution where I commented the
return.

Would you like me to give a try to your two suggestions?

**Comment by alfredh** [comment]
at least the changes in mwi module should be safe to apply
(feel free to commit directly to master).

the changes in ua.c requires a bit more thinking :)

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:
> at least the changes in mwi module should be safe to apply
> (feel free to commit directly to master).

OK, how about this mwi.c simplification:

```
static void tmr_handler(void *arg)
{
	struct le *le;

	(void)arg;

	for (le = list_head(uag_list()); le; le = le->next) {
		struct ua *ua = le->data;
		struct account *acc = ua_account(ua);

		if ((account_regint(acc) == 0) &&
		    (strcmp(account_mwi(acc), "yes") == 0)) {
			mwi_subscribe(ua);
		}
	}
}
```

=>

```
static void tmr_handler(void *arg)
{
	struct le *le;

	(void)arg;

	for (le = list_head(&mwil); le; le = le->next) {
		struct mwi *mwi = le->data;
		mwi_subscribe(mwi->ua);
	}
}
```

Has worked OK in my tests.

**Comment by juha-h** [comment]
I committed the mwi simplifications.  Now it remains to figure out the best way to modify ua_stop_all() so that both mwi and non-mwi UAs will get stopped.

**Comment by alfredh** [comment]
great, thanks a lot.

I think we should just check the ua ref-count, and decide on that what to do.

if the ua ref-count is 1, it means that no-one outside ua.c has a reference to it,
so it is safe to destroy it right away.

it should also disconnect any active calls, and remove itself from the list,
in all cases.

so I think something like this:

```c
	/* check if someone else has grabbed a ref to ua */
	le = uag.ual.head;
	while (le) {

		struct ua *ua = le->data;
		le = le->next;

		ua_event(ua, UA_EVENT_SHUTDOWN, NULL, NULL);

		if (mem_nrefs(ua) > 1) {
			++ext_ref;
		}

		list_unlink(&ua->le);
		list_flush(&ua->calls);
		mem_deref(ua);
	}
```

**Comment by juha-h** [comment]
How about https://github.com/alfredh/baresip/tree/ua_stop_all?

**Comment by alfredh** [comment]
many thanks ! 

merged to master in 97c6625245f994d43ed0b41e6c090b1ee6850f94
